book/src/persistent-account-storage.md

```diff
@@ -0,0 +1,148 @@
+# Persistent Account Storage
+
+The set of Accounts represent the current computed state of all the transactions
+that have been processed by a fullnode.  Each fullnode needs to maintain this
+entire set.  Each block that is proposed by the network represents a change to
+this set, and since each block is a potential rollback point the changes need to
+be reversible.
+
+Persistent storage like NVMEs are 20 to 40 times cheaper than DDR.  The problem
+with persistent storage is that write and read performance is much slower then
+DDR, and care must be taken in how data is read or written to.  Both reads and
+writes can be split between multiple storage drives and accessed in parallel.
+This design proposes a data structure that allows for concurrent reads and
+concurrent writes of storage.   Writes are optimized by using an AppendVec data
+structure, which allows a single writer to append while allowing access to many
+concurrent readers.  The accounts index maintains a pointer to a spot where the
+account was appended to for every fork, thus removing the need for explicit
+check-pointing of state.
+
+# AppendVec
+
+AppendVec is a data structure that allows for random reads concurrent with a
+single append only writer.  Grow, or resizing the capacity of the AppendVec
+requires exclusive access.  This is implemented with an atomic `offset`, which
+is updated at the end of a completed append.
+
+The underlying memory for an AppendVec is a memory mapped file.  Memory mapped
+files allow for fast random access, and paging is handled by the OS.
+
+# Account Index
+
+The account index is designed to support a single index for all the currently
+forked Accounts.
+
+```
+type AppendVecId = usize;
+
+type Fork = u64;
+
+struct AccountMap(Vec<Fork, (AppendVecId, u64)>);
+
+type AccountIndex = HashMap<Pubkey, AccountMap>;
+
+```
+
+The index is a map of account Pubkeys to a map of forks and the location of the
+Account data in an AppendVec.  To get the latest version of an account:
+
+```
+/// Load the account for the pubkey.
+/// This function will load the account from the greatest or equal to fork. 
+/// * forks - A vector of forks indicating a priority list of forks. The earlier
+///   forks are loaded first if available.
+/// * pubkey - The Account's public key.
+pub fn load_slow(&self, forks: &[u64], pubkey: &Pubkey) -> Option<&Account>
```

Comment:
this is out of date since we started new_from_parent(), wherein parenting is now inside accounts_db

Reply:
Can you push an update?

Reply:
you need to give us permission, or I can push a separate PR...

Reply:
#3076